sql/supabase_vault--0.3.0.sql

```diff
@@ -0,0 +1,123 @@
+CREATE OR REPLACE FUNCTION vault._crypto_aead_det_encrypt(message bytea, additional bytea, key_id bigint, context bytea = 'pgsodium', nonce bytea = NULL)
+RETURNS bytea
+AS 'MODULE_PATHNAME', 'pgsodium_crypto_aead_det_encrypt_by_id'
+LANGUAGE c IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION vault._crypto_aead_det_decrypt(message bytea, additional bytea, key_id bigint, context bytea = 'pgsodium', nonce bytea = NULL)
+RETURNS bytea
+AS 'MODULE_PATHNAME', 'pgsodium_crypto_aead_det_decrypt_by_id'
+LANGUAGE c IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION vault._crypto_aead_det_noncegen()
+RETURNS bytea
+AS 'MODULE_PATHNAME', 'pgsodium_crypto_aead_det_noncegen'
+LANGUAGE c IMMUTABLE;
+
+CREATE TABLE vault.secrets (
+  id uuid     PRIMARY KEY DEFAULT gen_random_uuid(),
+  name        text,
+  description text NOT NULL default '',
+  secret      text NOT NULL,
+  key_id      uuid,
+  nonce       bytea DEFAULT vault._crypto_aead_det_noncegen(),
+  created_at  timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at  timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP
+);
+
+COMMENT ON TABLE vault.secrets IS 'Table with encrypted `secret` column for storing sensitive information on disk.';
+
+CREATE UNIQUE INDEX ON vault.secrets USING btree (name) WHERE name IS NOT NULL;
+
+DROP VIEW IF EXISTS vault.decrypted_secrets;
+CREATE VIEW vault.decrypted_secrets AS
+SELECT s.id,
+  s.name,
+  s.description,
+  s.secret,
+  convert_from(
+    vault._crypto_aead_det_decrypt(
+      message := decode(s.secret, 'base64'::text),
+      additional := convert_to(s.id::text, 'utf8'),
+      key_id := 0,
+      context := 'pgsodium'::bytea,
+      nonce := s.nonce
+    ),
+    'utf8'::name
+  ) AS decrypted_secret,
+  s.key_id,
+  s.nonce,
+  s.created_at,
+  s.updated_at
+FROM vault.secrets s;
+
+GRANT ALL ON SCHEMA vault TO pgsodium_keyiduser;
+GRANT ALL ON TABLE vault.secrets TO pgsodium_keyiduser;
+GRANT ALL ON vault.decrypted_secrets TO pgsodium_keyiduser;
```

Comment:
these roles come from pg_sodium. for new installs on 0.3.0 is pg_sodium still required?

there is no exception in tests currently due to fixtures.sql creating these roles but wouldn't that need to be directly in this setup script for prod?

Reply:
The roles will stay when pgsodium is dropped (Postgres doesn't track extension dependents for roles).

I'll see if I can change how the permissions is set up so we can further remove pgsodium-isms, but at worst we just create these roles at AMI creation once we stop creating pgsodium by default.